### ISSUE TITLE
Formula should be adjusted

### ISSUE THREAD

**Issue by TwiN** [thecsw/memeinvestor_bot#74]
**NOTE:** I'm not saying that there is anything wrong with investing early! However, if a meme ends up with 10 upvotes at the end of the 4 hours mark, the meme should not be considered as high quality, thus the value of the meme on the market should yield a lower investment return than a meme that has 500+ upvotes after 4 hours.

# Problem

Making MemeCoins is too easy with the current formula, and it needs to be adjusted. Even if a meme is bad, as long as one invests as early as possible, he's **guaranteed** to make a fortune.

The only factor currently taken in consideration is the submission karma when the user invests and the karma when the 4 hours passed. **The value of the meme itself on the market is completely ignored**.

## How?

The key is to invest as early as possible, **it almost always doesn't matter if the meme is good or not**.

### Procedure 

- New meme is published (karma = 1).
- User invests 1000 MemeCoins as quickly as possible.
- User waits for bot to confirm the investment.
- Bot confirms that 1000 MemeCoins have been invested when the submission had 1 karma.
- User upvotes the meme

At this point, the meme already has 2 karma, which bypasses the 1.2 threshold. While it can get downvoted, it is much more likely that at the very least** a few people will upvote, even if the meme is not necessarily high quality. 

Maybe you don't see the problem yet, so to be more clear, if after 4 hours, the meme has a score of 10, the meme should be considered bad (an average of 2.5 upvotes per hour is bad). However, the user who voted when there was 1 upvote will still make a ton of MemeCoins **because every single upvote is a multiple of initial investment**, (10 upvotes = 10 x 1 upvotes...)

### Some numbers

| Investment | Start karma | End karma | result |
| ------------- | ------------- | ----------- | ------ |
| 1000 | 1 | 5 | 1710 | 
| 1000 | 1 | 10 | 2154 | 
| 1000000000 | 1 | 5 | 1709975947 | 
| 1000000000 | 1 | 10 | 2154434690 | 

## Solution

To be as accurate as possible, we'd need to monitor the rise of each memes over time, and calculate the investment return based on the growth factor over time, but clearly, that's not possible due to the large amount of submissions. 

Instead, a range threshold should also be added.

With real stocks, we'd have a volume and a price per share, however, there's no volume and no price per share for memes. To replace the volume, the total amount of upvotes should be used.

For instance:

- `End karma < 10`: max investment return = 120%
- `End karma < 20`: max investment return = 140%
...
- `End karma > 100`: no limit to max investment return

By implementing a range like that, early investors will not get a big investment return if after 4 hours, the meme has a small volume on the market whereas if the meme has a large volume, the user would not be affected by this limitation. 

# TL;DR

After 4 hours, if the meme has...
- A low total score, limit the maximum investment return.
- A high total score, do not limit the maximum investment return.

I'm a little sleepy, so if there's anything unclear, let me know, I'll clarify.

**Comment by ometrist** [comment]
> At this point, the meme already has 2 karma, which bypasses the 1.2 threshold

The threshold is ~1.7x growth in upvotes which is still met by going from 1 to 2 upvotes, just wanna clarify. I highly agree something needs to change. **I LIKE YOUR SOLUTION!** I'd change the fact that a meme with 100+ upvotes would have a maximum return of 3800%, similar to what the max multiplier is currently set at. 

How easy would something like this be to code?

**Comment by jimbobur** [comment]
I've had something in the works to address this exact problem. I've been tinkering with it since Sunday. 
Also note, the relative change in upvotes is cube rooted, so a 1.2 threshold on the factor is actually a 1.728 threshold on the relative change in upvotes, but that's by the by anyway.

**Comment by jimbobur** [comment]
So basically, as you've correctly observed, there needs to be some range thresholding on lower final upvotes. This can be achieved by multiplying the current cuberoot function by a sigmoid (logistic) function of the final post upvotes after 4h with a sensibly-chosen midpoint and steepness parameter. To this end, I've been playing around with this numerically, and have also coded it in a standalone experimental version of the calculate function.  
Below is a plot of the original final upvotes needed to reach the win/lose (blue) and lose/nothing (red) thresholds (dotted straight lines) along with the same final upvote thresholds for the new cuberoot * sigmoid version (solid lines).
![image](https://user-images.githubusercontent.com/34943023/40967963-54d88bb6-68ac-11e8-8ba6-12e824992975.png)
The shape/height/region-of-effect of this sigmoid is tunable as desired by changing the steepness and midpoint parameters of the function. Here, for example, the midpoint is set at 500 and the steepness is 6(1/500).  
  
The beauty of using a sigmoid function is it achieves what we want in terms of being able to ramp up the win threshold for lower finishing upvotes whilst not affecting higher finishing upvotes. Below are the same numbers from the above plot put into a table to illustrate what I mean:


Starting Upvotes | Original CUBEROOT Win/Lose threshold final upvotes | CUBEROOT * SIGMOID      Win/Lose threshold final upvotes
-- | -- | --
1 | 1.728 | 366.4896402
2 | 3.456 | 388.2039261
3 | 5.184 | 401.4601651
4 | 6.912 | 411.1689456
5 | 8.64 | 418.8996844
10 | 17.28 | 444.2845297
100 | 172.8 | 561.0339549
300 | 518.4 | 691.3031407
600 | 1036.8 | 1041.49311
1000 | 1728 | 1728.002065
1500 | 2592 | 2592
2000 | 3456 | 3456
2500 | 4320 | 4320
5000 | 8640 | 8640

As you can see, at higher finishing upvotes (far from the 500 midpoint) the sigmoid function is very close to 1, so multiplying by it has no effect. In the graph, this can be seen as the solid line (original * sigmoid) asymptotes to the dotted line (original). However, at lower finishing upvotes, it has the effect of smoothly ramping up the difficulty in winning on an investment.

**Comment by ometrist** [comment]
I like the work you're putting into this and I'm glad to see you were doing it even before anyone finally brought it up on here. Your graph looks very nice and the table makes it easy to understand. 

In my honest investing opinion, it looks like based on those numbers that people would lose on almost every investment. I'd be afraid to invest in anything. This nerfs /new posts, but by too much and still doesn't reward /hot posts enough to counteract.

 I feel like if we lowered the win/lose threshold somehow, it would be better. People like winning, whether it's a little or a lot!

**Comment by jimbobur** [comment]
As I said, the above numbers/plot are just an example; the function parameters I chose were completely arbitrary (whilst still being something at least vaguely sensible). The form of the sigmoid function is tunable as desired by adjusting its slope/midpoint (/max value too, if needed).

**Comment by jimbobur** [comment]
To illustrate this, I've just re-ran the numbers. I set the midpoint to 250 and solved numerically for the sigmoid steepness parameter needed to set the final upvotes win/lose threshold for starting at 0 upvotes to be (as an example) 150 upvotes. The required steepness is 0.0123190233003174 and the plot now looks like this:  
![image](https://user-images.githubusercontent.com/34943023/40971752-8c87fa42-68b6-11e8-9c70-69978b1c808a.png)

and the table is now:  

Starting Upvotes | Original CUBEROOT      Win/Lose threshold final upvotes | CUBEROOT * SIGMOID      Win/Lose threshold final upvotes
-- | -- | --
0 | 1.728 | 150
1 | 1.728 | 150
2 | 3.456 | 170.3786075
3 | 5.184 | 183.068106
4 | 6.912 | 192.4823059
5 | 8.64 | 200.0528809
10 | 17.28 | 225.3998037
100 | 172.8 | 355.6730353
300 | 518.4 | 555.3836952
600 | 1036.8 | 1036.99159
1000 | 1728 | 1728.000064
1500 | 2592 | 2592
2000 | 3456 | 3456
2500 | 4320 | 4320
5000 | 8640 | 8640

So the function can be adjusted to produce whatever level of effect we want/seems reasonable and over whatever range of upvote values we want to affect. We just need to decide what we consider 'fair' to investors.

**Comment by ometrist** [comment]
> function is tunable as desired 

i read that as "function is turntable and desired" twice. it's past 4AM for me here, you'll have to forgive me. 

You use a lot of words i do know and a lot of words I don't know lol.  If I look at the exact function written out that you used to make the "CUBEROOT * SIGMOID Win/Lose threshold final upvotes" column on your second table, I would love to play with it. I took calculus and stats *years* ago, but I have a decent understanding. 

Is it possible for you to show me the function?

**Comment by jimbobur** [comment]
> it's past 4AM for me here, you'll have to forgive me.  

Blimey, get some rest, you must be knackered!

The full function is the old cuberoot[relative upvote change] function multiplied by a (logistic) sigmoid function of the final upvotes, so, writing it all out on one line:
  **For u0>0**  
`M = [cuberoot function] * [sigmoid function] `
**i.e.**
` M = [relative upvote change]^(1/3) * [sigmoid function] `  
**SO**:
 ` M = [ 1+(u-u0)/u0 ]^(1/3) * [1/( 1+e^(-k(u-m) ) ]` 
**For u0=0** (special case: change relative diff calculation to avoid DIV0 error)
`M = [ u ]^(1/3) * [1/( 1+e^(-k(u-m) ) ]`  
  **where**:
M = investment return factor
u0 = initial upvotes
u = final upvotes
k = steepness of sigmoid
m = sigmoid midpoint (sigmoid = 0.5 at this value of u)

NOTE: As far as I can tell this function is not analytically soluble for u, even in the 'simpler' u0=0 case, so if you want to find the final upvotes required to reach the M=1.2 threshold for a given set of sigmoid function parameters **k,m**, you'll need to solve for u numerically.

**Comment by thecsw** [comment]
You know, I really like this side project. We talk so seriously, use *complex* math and draw some sophisticated graphs to represent a growth in popularity of **memes**. Let this project rise and continue! Thank you for your time and contribution!

**Comment by TwiN** [comment]
@jimbobur Sounds great! I knew you could find a better solution than me for that, that's why I made an issue out of this instead of making a PR directly. 

However, by doing this, we're making the gap between the top 1% investors and the other 99% almost impossible to close. Maybe we should consider making a poll on r/MemeEconomy about whether we should reset the economy? 

Unlike a real economy, we have the option of resetting it, but doing so often would piss people off. Seeing how there's already a few people with more than 1 trillion MemeCoins, we clearly need to do something about this.

@ometrist go to bed, lmfao

**Comment by ometrist** [comment]
@jimbobur 

I've tried messing with the formula with no success better than what you've presented to me already, I'm gonna take myself out of the equation in creating a new investment return function. If you ever need an eye exam you let me know though, that's my area of expertise 👓 👀

**Comment by thecsw** [comment]
This formula is a bit outdated now. Similar issue #84 can help you. Thanks.
